Export JVM metrics from the javaagent using micrometer

mateuszrzeszutek opened the pull request:
README changes will be added in a separate PR

jkwatson commented on custom/src/main/java/com/splunk/opentelemetry/micrometer/OtelNamingConvention.java:
```diff
@@ -0,0 +1,56 @@
+ *       all metrics with {@code jvm.}
+ * </ol>
+ */
+class OtelNamingConvention implements NamingConvention {
```
I don't see any unit tests for this class. I think that would be good to have.

mateuszrzeszutek replied:
Definitely -- I must have overlooked this class, I was so focused on making gradle work 😄

jkwatson reviewed (approved):
This look really great. Just would like some unit tests for `OtelNamingConvention`, but that could come in a separate PR if you would prefer.

breedx-splk commented on bootstrap/src/main/java/com/splunk/opentelemetry/javaagent/bootstrap/GlobalMetricsTags.java:
```diff
@@ -0,0 +1,49 @@
+ * This class holds the list of tags that should be added to all meters registered by the javaagent.
+ * It needs to be loaded by the bootstrap classloader to be accessible inside instrumentations.
+ */
+public final class GlobalMetricsTags {
```
I don't want this thought to get in the way of this PR...but I would love for us to also consider ways of trying to avoid this static/global pattern down the road...even if it means passing the tags into all instrumentations that need it.  Or something.  I dunno....I just worry about this being hard to maintain.

breedx-splk replied:
I also realize that passing it into an AutoService may be difficult/impossible...but still.  😁

mateuszrzeszutek replied:
I first looked into setting global tags in the meter registry itself, which would be a much cleaner solution -- no need to remember about passing global tags to your new meter, the registry would it automagically. Unfortunately it required digging really deep into `MeterRegistry` internals and overriding >10 different methods to make this work 🥶  

And the fact that those tags need to be accessible both in the agent and application classloaders makes "fancy" (non-global variable 😂 ) implementation much harder...

breedx-splk replied:
Yeah, of course.  Seems reasonable for now then.

breedx-splk commented on custom/src/main/java/com/splunk/opentelemetry/micrometer/MicrometerInstaller.java:
```diff
@@ -0,0 +1,73 @@
+    if (runtime != null) {
+      globalTags.add(Tag.of("runtime", runtime));
+    }
+    if (pid != null) {
```
Would be good to have unit test coverage on this stuff too.

kubawach replied:
I think that this (meaning - creating a list of tags) could even be turned into a builder attached to the GlobalMetricsTags class. This suggestion is a bit in line with @breedx-splk previous idea of moving most of the code there which could not be implemented due to class loading issue.

mateuszrzeszutek replied:
I decided to keep the `GlobalTagsBuilder` class in the `custom` module so that it can use the OTel Resource API.

breedx-splk commented on custom/src/main/java/com/splunk/opentelemetry/micrometer/OtelNamingConvention.java:
```diff
@@ -0,0 +1,56 @@
+  @Override
+  public String name(String name, Meter.Type type, String baseUnit) {
+    String metricName = signalfxNamingConvention.name(name, type, baseUnit);
+    return metricName.startsWith("jvm.") ? "runtime." + metricName : metricName;
```
Unit tests please!

breedx-splk commented on custom/src/main/java/com/splunk/opentelemetry/micrometer/SplunkMetricsConfig.java:
```diff
@@ -0,0 +1,67 @@
+class SplunkMetricsConfig implements SignalFxConfig {
+  @Override
+  public boolean enabled() {
+    return Config.get().getBooleanProperty("splunk.metrics.enabled", true);
```
Let's limit the number of global/static calls and create a config field on this class that is passed into the constructor.  Same with `Resource` (below).

breedx-splk commented on custom/src/main/java/com/splunk/opentelemetry/micrometer/SplunkMetricsConfig.java:
```diff
@@ -0,0 +1,67 @@
+  public String uri() {
+    // right now the default value points to SmartAgent endpoint
+    return Config.get()
+        .getProperty("splunk.metrics.endpoint", "http://localhost:9080/v2/datapoint");
```
Promote the value to a constant to raise visibility/findability?

breedx-splk commented on custom/src/main/java/com/splunk/opentelemetry/micrometer/SplunkMetricsConfig.java:
```diff
@@ -0,0 +1,67 @@
+  public Duration step() {
+    long stepMillis =
+        Long.parseLong(Config.get().getProperty("splunk.metrics.export.interval", "30000"));
+    return Duration.ofMillis(stepMillis);
```
Would like to see unit test coverage on this.

breedx-splk commented on instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/jvmmetrics/JvmMetricsInstaller.java:
```diff
@@ -0,0 +1,50 @@
+    new ClassLoaderMetrics(tags).bindTo(Metrics.globalRegistry);
+    new JvmGcMetrics(tags).bindTo(Metrics.globalRegistry);
+    new JvmMemoryMetrics(tags).bindTo(Metrics.globalRegistry);
+    new JvmThreadMetrics(tags).bindTo(Metrics.globalRegistry);
```
👍

breedx-splk reviewed (changes requested):
This is awesome stuff!  Had a few ideas/comments, bu t looks good overall to me.

mateuszrzeszutek commented on custom/src/main/java/com/splunk/opentelemetry/micrometer/MicrometerInstaller.java:
```diff
@@ -0,0 +1,73 @@
+    Metrics.addRegistry(createSplunkMeterRegistry());
+  }
+
+  private static List<Tag> getGlobalTags() {
```
Unfortunately `GlobalMetricsTags` is in the bootstrap classloader and has no access to the otel SDK - I had to do it in this class (which is in the agent classloader and thus can see SDK classes)

breedx-splk replied:
Awww shucks! I didn't consider that.  I guess this is a form of bridging....

iNikem commented on custom/src/main/java/com/splunk/opentelemetry/micrometer/SplunkMetricsConfig.java:
```diff
@@ -0,0 +1,67 @@
+  @Override
+  public Duration step() {
+    long stepMillis =
+        Long.parseLong(Config.get().getProperty("splunk.metrics.export.interval", "30000"));
```
Is this method called often or just once? In the former case we should extract this to a constant, I think.

mateuszrzeszutek replied:
Good catch - micrometer does not cache this at all (despite parsing being present in the base `PushRegistryConfig` class) and calls it every time a meter is created.

iNikem commented on gradle/instrumentation.gradle:
```diff
@@ -37,17 +36,20 @@ shadowJar {
+
     jvmArgs "-Dotel.javaagent.debug=true"
-    jvmArgs "-javaagent:${configurations.testAgent.files.first().absolutePath}"
+    jvmArgs "-javaagent:${project(":testing:agent-for-testing").tasks.shadowJar.archiveFile.get().asFile.absolutePath}"
```
I think you should change it back :) https://docs.gradle.org/current/userguide/cross_project_publications.html

mateuszrzeszutek replied:
I think I fixed it 😅

iNikem commented on instrumentation/jvm-metrics/src/test/java/com/splunk/opentelemetry/jvmmetrics/JvmMetricsTest.java:
```diff
@@ -0,0 +1,41 @@
+    var meterNames = TestMetricsAccess.getMeterNames();
+
+    // classloader metrics
+    assertTrue(meterNames.contains("jvm.classes.loaded"));
```
Looking at `OtelNamingConvention`, I would expect seeing `runtime` somewhere here. Why not?

mateuszrzeszutek replied:
Because `OtelNamingConvention` is not actually present in the test javaagent, we don't include `project(":custom")` there. In another PR I'd like to add metrics verification to smoke tests (once I figure out how...) so the full name will be checked there.

kubawach commented on custom/src/main/java/com/splunk/opentelemetry/micrometer/OtelNamingConvention.java:
```diff
@@ -0,0 +1,56 @@
+ * </ol>
+ */
+class OtelNamingConvention implements NamingConvention {
+  private final NamingConvention signalfxNamingConvention;
```
I'd cut "signalfx" prefix here (does not seem to be now-defunct-signalfx related).

kubawach commented on custom/src/main/java/com/splunk/opentelemetry/micrometer/MicrometerInstaller.java:
```diff
@@ -0,0 +1,73 @@
+  private static SignalFxMeterRegistry createSplunkMeterRegistry() {
+    SignalFxMeterRegistry signalFxRegistry =
+        new SignalFxMeterRegistry(new SplunkMetricsConfig(), Clock.SYSTEM);
+    NamingConvention signalFxNamingConvention = signalFxRegistry.config().namingConvention();
```
Just wondering - is it possible to get a null naming conv here?

mateuszrzeszutek replied:
No -- `SignalFxMeterRegistry` always sets its own convention.

kubawach reviewed (approved):
Some minor comments.

mateuszrzeszutek commented on custom/build.gradle:
```diff
@@ -10,15 +10,23 @@ def relocatePackages = ext.relocatePackages
 dependencies {
   compileOnly(project(":bootstrap"))
-  implementation("io.opentelemetry:opentelemetry-sdk:${versions["opentelemetry"]}")
+  implementation("io.opentelemetry:opentelemetry-sdk:${versions.opentelemetry}")
   implementation("io.opentelemetry:opentelemetry-sdk-extension-autoconfigure:${versions.opentelemetryAlpha}")
+  implementation("io.opentelemetry:opentelemetry-semconv:${versions.opentelemetryAlpha}")
   implementation("io.opentelemetry:opentelemetry-exporter-jaeger-thrift:${versions.opentelemetry}")
+  implementation("io.opentelemetry.instrumentation:opentelemetry-instrumentation-api:${versions.opentelemetryJavaagentAlpha}")
   implementation("io.opentelemetry.javaagent:opentelemetry-javaagent-spi:${versions.opentelemetryJavaagentAlpha}")
```
🤔 
Most of them, probably -- I'll change that.

mateuszrzeszutek commented:
Seems like the tests have passed (at least the linux ones...) -- @iNikem @kubawach @breedx-splk @jkwatson I've addressed all your comments, please re-review this PR 🙏

jkwatson commented on custom/src/test/java/com/splunk/opentelemetry/micrometer/OtelNamingConventionTest.java:
```diff
@@ -0,0 +1,68 @@
+class OtelNamingConventionTest {
+  @Mock NamingConvention namingConventionMock;
+
+  NamingConvention otelNamingConvention;
+
+  @BeforeEach
+  void setUp() {
+    otelNamingConvention = new OtelNamingConvention(namingConventionMock);
+  }
```
Putting the class under test here I think is less readable than putting it in each of the test cases. and, it would be fewer lines of code to just put this in each of the test cases, as well.

jkwatson commented on custom/src/test/java/com/splunk/opentelemetry/micrometer/GlobalTagsBuilderTest.java:
```diff
@@ -0,0 +1,66 @@
+class GlobalTagsBuilderTest {
+  @Test
+  void shouldBuildEmptyTagsList() {
+    var resource = Resource.create(Attributes.empty());
```
I think you lost a `// given` here.

jkwatson commented on custom/src/test/java/com/splunk/opentelemetry/micrometer/GlobalTagsBuilderTest.java:
```diff
@@ -0,0 +1,67 @@
+                12345L));
+
+    // when
+    var tags = new GlobalTagsBuilder(resource).build();
```
java 11 in tests.. is there gradle magic that makes sure our code doesn't start using java 11 APIs that would break users on java 8?
